events/entity_counts.go

```diff
@@ -0,0 +1,97 @@
+package events
+
+import (
+	"fmt"
+	"github.com/openziti/edge/controller/persistence"
+	"github.com/openziti/foundation/util/cowslice"
+	"github.com/pkg/errors"
+	"go.etcd.io/bbolt"
+	"reflect"
+	"time"
+)
+
+const EntityCountEventNS = "edge.entityCounts"
+
+type EntityCountEvent struct {
+	Namespace string           `json:"namespace"`
+	Timestamp time.Time        `json:"timestamp"`
+	Counts    map[string]int64 `json:"counts"`
+	Error     string           `json:"error"`
+}
+
+func (event *EntityCountEvent) String() string {
+	return fmt.Sprintf("%v timestamp=%v counts=%+v err=%v",
+		event.Namespace, event.Timestamp, event.Counts, event.Error)
+}
+
+var entityCountEventGenerator func(interval time.Duration)
+
+func generateEntityCountEvent(dbProvider persistence.DbProvider, stores *persistence.Stores) {
+	handlers := getEntityCountEventHandlers()
+	if len(handlers) == 0 {
+		return
+	}
+
+	event := &EntityCountEvent{
+		Namespace: EntityCountEventNS,
+		Timestamp: time.Now(),
+		Counts:    map[string]int64{},
+	}
+
+	for _, store := range stores.GetStoreList() {
+		err := dbProvider.GetDb().View(func(tx *bbolt.Tx) error {
+			_, count, err := store.QueryIds(tx, "true limit 1")
+			if err != nil {
+				return err
+			}
+			event.Counts[store.GetEntityType()] = count
+			return nil
+		})
+
+		if err != nil {
+			event.Error = err.Error()
+			break
+		}
+	}
+
+	for _, handler := range handlers {
+		handler.AcceptEntityCountEvent(event)
+	}
+}
+
+type EntityCountEventHandler interface {
+	AcceptEntityCountEvent(event *EntityCountEvent)
+}
+
+var entityCountEventHandlerRegistry = cowslice.NewCowSlice(make([]EntityCountEventHandler, 0))
+
+func getEntityCountEventHandlers() []EntityCountEventHandler {
+	return entityCountEventHandlerRegistry.Value().([]EntityCountEventHandler)
+}
+
+func registerEntityCountEventHandler(val interface{}, config map[interface{}]interface{}) error {
```

Comment:
Why take in `val` then type assert it to `EntityCountEventHandler` instead of defining the function as `val EntityCountEventHandler` ?

Reply:
The registration function has to conform to 

`type registrationHandler func(handler interface{}, config map[interface{}]interface{}) error`

This generic type is used by all the event types, so it can't be typed to a specific one.